src/useAnswersState.tsx

```diff
@@ -1,15 +1,76 @@
-import { useContext, useRef, useState } from 'react';
+import React, { useCallback, useContext, useEffect, useRef, useState } from 'react';
 import { State } from '@yext/answers-headless';
 import { AnswersHeadlessContext } from './AnswersHeadlessContext';
 import useLayoutEffect from 'use-isomorphic-layout-effect';
+import { useSyncExternalStoreWithSelector } from 'use-sync-external-store/shim/with-selector';
 
 export type StateSelector<T> = (s: State) => T;
 
+const canUseDOM = !!(typeof window !== 'undefined'
+  && typeof window.document !== 'undefined'
+  && typeof window.document.createElement !== 'undefined');
+const isServerEnvironment = !canUseDOM;
+
+/**
+ * use-sync-external-store/shim does not support getServerSnapshot for pre-18 React versions.
+ * It requires user of the shim to handle subscription logic and return the correct snapshot value.
+ * As such, server side rendering from pre-18 React version will use a version of useAnswersState that
+ * manually handle memoization of state selector and register subscription without useSyncExternalStore call.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const useAnswersState = (React as any).useSyncExternalStore === undefined && isServerEnvironment
+  ? useAnswersStateWithManualStoreSync : useAnswersStateWithReactStoreSync;
```

Comment:
so `useAnswersStateWithManualStoreSync` is only for react 17 SSR?

Reply:
does `useAnswersStateWithManualStoreSync` work with react 18 as well?

Reply:
I'm also kind of curious if you read anything about what useSyncExternalStore is doing under the hood? If not then nbd

Reply:
When I was testing with `startTransition` to try to simulate tearing issues, `useAnswersStateWithManualStoreSync` doesn't work well with concurrent feature in react 18. 
Yup! `useAnswersStateWithManualStoreSync` would be for react 17 SSR only since useSyncExternalStore can handle the client side rendering for 17 and automatically use the builtin useSyncExternalStore API for 18. 
I didn't dig too much into what useSyncExternalStore does internally but it looks like it is a [reform](https://github.com/reactwg/react-18/discussions/86) of [useMutableSource](https://github.com/reactwg/react-18/discussions/84) if you are familiar with that